### ISSUE TITLE
Softer handling of expired locks

### ISSUE THREAD

**Issue by matthijskooijman** [rparent/django-lock-tokens#6]
AFAIU, django-lock-tokens currently exposes the following interface
 1. You can obtain a lock on any object, provided that no other lock is active
 2. When a lock is not renewed for some time, it expires and effectively disappears (note that it remains in the database to be [deleted when a new lock is taken](https://github.com/rparent/django-lock-tokens/blob/ddb5dbc7ae5bf19d361604ffdf3c52c6d8129a1a/lock_tokens/models.py#L63-L65), but it is already [hidden from most operations](https://github.com/rparent/django-lock-tokens/blob/ddb5dbc7ae5bf19d361604ffdf3c52c6d8129a1a/lock_tokens/managers.py#L13), so things behave as if it was deleted already).
 3. When you want to change an object, no active lock by someone else must be active (so either you have an active lock, or no lock is active).

I wonder if the way expired locks are handled is the best approach here. In particular, consider this case:
 1. Alice takes a lock on an object and opens up its editing interface.
 2. Alice walks away from her computer, the lock expires.
 3. Alice returns, clicks save. She now no longer has a valid lock.

In this case, Alice cannot renew its existing lock, since it has effectively disappeared. However, with the current implementation of lock checking, she can still save her changes, since there will be no lock active, which `check_and_get_lock_token()` [allows, but with a warning](https://github.com/rparent/django-lock-tokens/blob/ddb5dbc7ae5bf19d361604ffdf3c52c6d8129a1a/lock_tokens/models.py#L89-L90)

However, now consider the case where, while Alice is gone, Bob takes a lock, makes some edit, saves the object and releases the lock again. When Alice returns, Bob's lock is gone again, so she can still save here pending changes (again, with a warning). However, this time, Alice's changes will *overwrite* Bob's changes, which I think is precisely what this locking mechanism should prevent.

The cause for this overwriting is that `check_lock_token` (and thus also `check_for_session`) returns `True` when no active lock exists. To really prevent this, it should only return true when the given lock (or the lock from the session) that was taken when the editing started, is still valid and active. If not, some other user *might* have changed the object.

However, if you would implement this, the "hard" handling of locking (effectively deleting them as soon as they expire) becomes somewhat problematic: If a user lets his lock expire, he can no longer save his changes, even when no other user has made changes in the meanwhile, in which case saving would not be problematic. I would suggest that expired locks are handled a bit more softly.

Here, you can distinguish two cases:
 - While the lock was expired, (no) changes were made to the object. If no changes were made since the lock was opened (rather, since the editing interface was loaded), it is safe to allow committing unsaved changes without risk of overwriting other changes. However, detecting whether an object changed is not currently within scope of the django-lock-tokens package (it is more in the area of optimistic locking), so this case is hard to handle inside the package.
 - While the lock was expired (no) locks were taken on the object. If no lock was taken, you can be sure that no changes were made and thus committing pending changes is safe. When a lock was taken, changes *might* have been made, so the safe approach is to deny committing changes.

I can imagine an implementation, where:
 - An expired lock remains valid and can be renewed and used when saving normally.
 - An expired lock becomes invalid (and is deleted) as soon as another lock is taken on the same object.
 - To save an object, a valid lock is always needed (so *no lock* is no longer acceptable).

I believe that this could be implemented with minimal changes to the current code, since expired locks are already kept around in the database.

If the current hard approach is deemed useful as well (or you want to keep more compatibility), you could also consider introducing a soft and hard expiration time. When the soft expiration time is reached, a token remains valid, but can be deleted when another lock is taken. When the hard expiration time is reached, a token becomes invalid immediately. Either or both of these times can be set to `None` to disable that expire time.

Similarly, changing the interpretation of no lock from valid (with warning logged) to invalid might not be appropriate for all usecases, but this might be a matter of adding an optional keyword argument (e.g. `missing_ok=True`) or global configuration option.

For the first case above (e.g. allow renewing a lock when another user has taken a lock in the meanwhile, but has not actually saved any data), I said that this is out of scope for this package and is more in the area of optimistic locking libraries that save a version number for each object. However, I later realized a possible implementation that could be within scope. You could change the implementatoin to allow (multiple) expired locks to co-exist with one active lock for the same object. Then, instead of deleting expired locks when a new lock is taken, you keep them around but delete them only when saving the object (i.e. when unlocking, you would indicate whether the lock was actually "used" or just released again, and when it was used, you delete previous expired locks). This means that an expired lock stays around, and can be renewed, until a change is made, after which all previous locks become unused. There might be some atomicity issues here (in particular, the current unique constraint cannot be kept) but I suspect those can be solved by using proper transactions. This does mean a bigger change to the current code, though.

**Comment by rparent** [comment]
At the moment, the only way to avoid the overwriting you describe is to "hold" the lock on the client side, for example by using the provided Javascript API client and [its `hold_lock` method](https://github.com/rparent/django-lock-tokens#locktokenshold_lockapp_label-model-object_id). 

Of course this is not satisfying, as something could go wrong on the js side when renewing lock - or the developer could forget to handle that part... 

I agree we should think of another way of preventing this case, handled on the server side. I have to think this through, but my first choice would be in favor of simply making the lock token mandatory when saving, which would prevent the case you describe in your example.

**Comment by rparent** [comment]
I have just made a try at preventing this issue. Feel free to review !

I implemented the minimal solution, which is to always deny committing changes when another lock has been taken, even if no changes were made. This way we are sure that overwriting changes is always prevented. 

The other suggested solution (i.e. allow committing changes if another lock has been taken but with no changes made) could be an interesting feature to be added, but I see it more as an improvement, whereas the changes I implemented are really a bug fix :)

**Comment by matthijskooijman** [comment]
> At the moment, the only way to avoid the overwriting you describe is to "hold" the lock on the client side, for example by using the provided Javascript API client and its hold_lock method.

Yeah, that's what @brendah was already doing, but we realized that this allows a lock to be held indefinitely when someone walks away from his computer. To prevent that, we thought about adding an "Are you still there?" popup, and stop renewing the lock until the user clicks yes (or perhaps autodetect user presence somehow and stop renewing after the user is gone for a while). However, when you then let the lock expire, you'll want some way for the user to renew his expired lock again, if possible, to not throw away work (the objects in Brenda's application are big rosters which might take some time to edit).


> I have just made a try at preventing this issue. Feel free to review !

Nice, thanks for the quick action :-D

> I implemented the minimal soluI implemented the minimal solution, which is to always deny committing changes when another lock has been taken, even if no changes were made. This way we are sure that overwriting changes is always prevented.tion, which is to always deny committing changes when another lock has been taken, even if no changes were made. This way we are sure that overwriting changes is always prevented.

That seems like a good first step, to make things more safe. I had a quick look over the changes, which look good at first glance (though some remarks below).

> The other suggested solution (i.e. allow committing changes if another lock has been taken but with no changes made) could be an interesting feature to be added, but I see it more as an improvement, whereas the changes I implemented are really a bug fix :)

It actually seems that you're sortof implementing this already. According to the docs, an expired lock is considered valid for `check_for_session()`, and at least in `save_model`, you try renewing the lock before saving (including an expired lock). I'm not entirely sure what [`lock_for_session`](https://github.com/rparent/django-lock-tokens/blob/eb9146e1b0aa8a8cd2f019bd97ce93c28b3e1d1f/lock_tokens/admin.py#L38) does when the session's lock no longer exists (e.g. in the unlikely event that another request locked the object, saved it and unlocked it again). I suspect it just creates a new lock, which means there's still a race in `save_model()` (though the changes in [`_lock()`](https://github.com/rparent/django-lock-tokens/commit/eb9146e1b0aa8a8cd2f019bd97ce93c28b3e1d1f#diff-079cebb3ce74098047bb3ea9ae303a90R81) could suggest the opposite). I guess the way to fix that is to (atomically?) check for a lock *and* renew it, rather than (possibly) taking a new lock?

Also, I'm not sure if `check_for_session()` actually returns `True` for expired locks, as documented. It seems this piece of code should handle that: https://github.com/rparent/django-lock-tokens/blob/eb9146e1b0aa8a8cd2f019bd97ce93c28b3e1d1f/lock_tokens/models.py#L105-L108

However, since `get_for_object()` already filters out expired locks: https://github.com/rparent/django-lock-tokens/blob/eb9146e1b0aa8a8cd2f019bd97ce93c28b3e1d1f/lock_tokens/managers.py#L11-L13

I think the above code is never ran for expired locks, but rather the `DoesNotExist` case is triggered, which returns False: https://github.com/rparent/django-lock-tokens/blob/eb9146e1b0aa8a8cd2f019bd97ce93c28b3e1d1f/lock_tokens/models.py#L102-L104

I haven't looked at the entirety of this change yet, I suspect there might be other corner cases in the API is it is presented or documented now (but I'm too tired to take in the big picture now :-p).

One omission in the docs I see now is that an "invalid token" is not currently defined anywhere.

**Comment by rparent** [comment]
> I'm not entirely sure what lock_for_session does when the session's lock no longer exists (e.g. in the unlikely event that another request locked the object, saved it and unlocked it again). I suspect it just creates a new lock, which means there's still a race in save_model() (though the changes in _lock() could suggest the opposite). I guess the way to fix that is to (atomically?) check for a lock and renew it, rather than (possibly) taking a new lock?

Currently `lock_for_session` raises an `InvalidToken` exception when the session's lock no longer exists, and renews it if it still exists (expired or not). This should prevent the race in `save_model`. By the way this exception is not handled in my last commit, I'm working on a fix commit.

I also need to handle the case where there is an invalid lock token in the session in the admin `change_view` method. In this specific case, we can just clear the session token for this object and try to create a new one, since this view will load the last version of the object from the database into the interface.

To sum up, it is ok to try to create a new token in `change_view`, but not in `save_model`: at this point user must hold a valid token in session to be able to save.

> I'm not sure if check_for_session() actually returns True for expired locks (...)since get_for_object() already filters out expired locks

You are right, I need to fix that.

**Comment by rparent** [comment]
Hello @matthijskooijman,

I just updated the README and I have taken the liberty of reusing your example in the first comment. I hope it is ok with you!

**Comment by matthijskooijman** [comment]
Yeah, feel free to use my comments as part of documentation, no problem.

The new section looks good (though one of the enumerations is not rendered properly, point 4 is underindented).

Also, I wonder if case 2 requires renewing the token before using it? I suspect that if it is not renewed, there is a race condition where you use an expired token while another request deletes it. I suspect the code already does this, but it might be good to make that explicit in the docs (probably also for case 1?).

**Comment by rparent** [comment]
Just updated README, thanks.

**Comment by BrendaH** [comment]
Hi there, finally (!!) got the time to work on this part of my project again.
Most issues seem to be fixed (THANKS!), however I have a javascript corner case that still has a soft-handling issue.

Timeout in settings.py is 120 secs.
I use a javascript `setTimeout` to prompt the user to actively renew his lock 60 seconds before the lock would actually expire. If he clicks ok before the lock expires then it is successfully renewed. However, if he waits too long there's a 404 error.  Even though nobody else took a lock on the item in the meantime and the lock is actually still there in the database (it's just over 2 minutes old).

The issue is in the `get_valid_lock_token_or_error` function in locktoken's view.py which still tries to get tokens with an expire time of within the last [timeout] seconds.
A token is actually passed to that function so it could check if it can find that certain token and if so just return it? (Though the function should probably be renamed)
The delete case where this function is also called could probably also work with this. There's no harm in deleting an expired token as far as I can tell. I'm not sure about the get though, that might actually expect a valid token to be returned.

**Comment by matthijskooijman** [comment]
I wonder if [`get_valid_lock_token_or_error`](https://github.com/rparent/django-lock-tokens/blob/master/lock_tokens/views.py#L38) should be using [`get_for_object`](https://github.com/rparent/django-lock-tokens/blob/master/lock_tokens/managers.py#L9) (or probably a new `get_for_contenttype_and_id` that can also be used by `get_for_object`)?
